mlflow/utils/file_utils.py

```diff
@@ -43,11 +44,54 @@
 from mlflow.utils import download_cloud_file_chunk
 from mlflow.utils.request_utils import download_chunk
 
-
 ENCODING = "utf-8"
 MAX_PARALLEL_DOWNLOAD_WORKERS = os.cpu_count() * 2
 
 
+class ArtifactProgressBar:
+    def __init__(self, desc, total, step, **kwargs) -> None:
+        self.total = total
+        self.step = step
+        self.pbar = None
+
+        if MLFLOW_ENABLE_ARTIFACTS_PROGRESS_BAR.get():
+            try:
+                from tqdm.auto import tqdm
+
+                self.pbar = tqdm(total=self.total, desc=desc, **kwargs)
+            except ImportError:
+                pass
+
+    @classmethod
+    def chunks(cls, file_size, desc, chunk_size):
+        return cls(
+            desc,
+            total=file_size,
+            step=chunk_size,
+            unit="iB",
+            unit_scale=True,
+            unit_divisor=1024,
+            miniters=1,
+        )
+
+    @classmethod
+    def files(cls, desc, total, step=1):
+        return cls(desc, total=total, step=step)
+
+    def update(self, index):
+        if self.pbar:
+            remaining = self.total - index * self.step
+            self.pbar.update(min(remaining, self.step))
+            self.pbar.refresh()
```

Comment:
Can we remove `index`? We can add an attribute representing the current process. If the diff between `total` and this attribute is smaller than the chunk size, we can update `total - self.<attr>` otherwise `self.step`.

```
self.<attr> += chunk_size
```

Reply:
This allows us to remove `enumerate`.

Reply:
Makes sense :)